### PR TITLE
FB Marketing Native: Insights Logarithmic Job Splitting

### PR DESCRIPTION
**Pull request by JustinASmith**
**Description:**

**source-facebook-marketing-native: Implement logarithmic job splitting for Facebook Insights**

This PR adds intelligent job splitting to handle Facebook's data limit errors, enabling successful syncs for large ad accounts that previously failed. This PR also introduces unit and integration tests that necessitated a good restructuring of the `tests/`. The prior `test_snapshots.py` are in `tests/e2e/` and produce the same snapshots as before.

This is a long Description, but worth the read IMO. The final _notes for  reviewers_ section briefly states how I tested this.

---

**Problem**

Facebook's Insights API returns `"Please reduce the amount of data"` errors (code 100, subcode 1487534) when query scope is too broad. Previously, this caused sync failures for accounts with many campaigns, adsets, or ads. Users with large accounts could not sync insights without the connector restarting frequently due to failed jobs.

---

**Solution**

Implement a job splitting system that automatically subdivides failed queries using logarithmic bisection through the ad hierarchy.

*Split decision tree:*

1. **ACCOUNT scope** — Discover all campaign IDs via insights query with 29-day lookback, then binary split into two CAMPAIGNS jobs
2. **Multiple entity IDs** — Binary split the ID list in half at the current scope level (e.g., 100 campaigns → two jobs of 50)
3. **Single CAMPAIGN** — Discover adset IDs using parent filter, then binary split into ADSETS jobs
4. **Single ADSET** — Discover ad IDs, then binary split into ADS jobs
5. **Single AD** — Cannot split further. Raise `CannotSplitFurtherError`, connector fails, cursor unchanged for retry on next run

*Key behaviors:*
- Each split job gets a fresh retry budget (3 attempts)
- Data limit errors (code 100, subcode 1487534) trigger *immediate* split without exhausting retries
- Results stream immediately as jobs complete via `asyncio.Queue`

---

**Leveraging Facebook's IN Filter Operator**

A key enabler of this architecture is Facebook's `filtering` parameter with the `IN` operator:

```python
filtering=[{
    "field": "campaign.id",
    "operator": "IN",
    "value": ["campaign_123", "campaign_456", "campaign_789"]
}]
```

This allows batching multiple entity IDs into a single job rather than submitting one job per entity.

---

**Efficiency Analysis**

The logarithmic approach is most efficient when smaller batches succeed after splitting. For an account with 1000 campaigns:

- **Best case:** Account-level job succeeds → 1 job total
- **Typical case:** Account fails, splits to 2 batches of 500, both succeed → 3 jobs total
- **Partial failures:** If one batch of 500 fails and needs further splitting → 5-10 jobs depending on where splitting stops

The approach becomes less efficient if failures cascade deeply. If *every* batch fails until reaching individual entities, the total job count approaches O(n). However, data limit errors typically occur because batches are too large—once batches are small enough, they succeed.

---

**Error Classification**

| Condition | Detection | Action | Retries |
|-----------|-----------|--------|---------|
| Job Failed | `async_status = "Job Failed"` | Retry → Split | Yes (3x) |
| Job Skipped | `async_status = "Job Skipped"` | Retry → Split | Yes (3x) |
| Job Timeout | Job running > 1 hour | Retry → Split | Yes (3x) |
| Data Limit | `error_code=100, error_subcode=1487534` | Split immediately | No |
| Atomic Failure | Single AD fails after retries | Raise Exception | N/A |

---

**Configuration**

```python
MAX_RETRIES = 3                           # Fresh budget per split job
MAX_CONCURRENT_JOBS = 100                 # Per-account semaphore limit
JOB_TIMEOUT = timedelta(hours=1)          # Max time per job
POLL_INTERVAL_SECONDS = 30                # Status check interval
ENTITY_ID_LOOKBACK_DAYS = 29              # 28-day attribution + 1 buffer
```

---

**Architecture**

The implementation lives in a new `insights/` module:

- **errors.py** — Custom exceptions (`CannotSplitFurtherError`, `DataLimitExceededError`) and `try_parse_facebook_api_error()` for converting HTTP responses to typed exceptions
- **types.py** — `InsightsJob` dataclass (scope, entity_ids, depth), `JobScope` enum, outcome types (`JobSuccess`/`JobSplit`), Pydantic models for API responses
- **metrics.py** — `JobMetrics` for observability (submitted/completed/split/failed counts, per-scope breakdown)
- **manager.py** — `FacebookInsightsJobManager` orchestrating concurrent execution

---

**Concurrency Model**

*Per-account job managers* — Each ad account gets its own `FacebookInsightsJobManager` instance with independent `asyncio.Semaphore`. This aligns with Facebook's per-account rate limiting—one account's workload cannot affect another's quota.

*Semaphore-based concurrency* — Jobs acquire the semaphore before execution and release on completion. The `_process_with_splitting()` method manages a job queue, launching tasks up to `MAX_CONCURRENT_JOBS` and processing results as they complete.

*Queue-based result streaming* — Results are pushed to an `asyncio.Queue` and yielded immediately rather than accumulating in memory. This keeps memory usage generally constant regardless of result set size since the CDK offloads documents to disk once we exceed a threshold without a checkpoint.

---

**Entity Discovery**

When splitting requires descending the hierarchy (e.g., account level → campaigns), we discover child entity IDs using a synchronous insights query:

- 29-day lookback window (28-day attribution + 1-day buffer)
- No breakdowns or time_increment (minimizes response size since we just care about IDs)
- Parent filter when descending (e.g., `campaign.id IN [...]` when finding adsets)

This finds entities with insights data in the attribution window, including paused or archived entities.

---

**Logging and Observability**

Jobs are logged with hierarchy context:

```
Executing ACCOUNT job [depth=0]
ACCOUNT job [depth=0] hit data limits, splitting immediately
Split ACCOUNT job -> 2 CAMPAIGNS jobs (1000 campaigns discovered, child_depth=1)
Executing campaigns job (500 entities) [depth=1, from ACCOUNT]
```

Final summary:

```
Job processing complete: submitted=5, completed=4, split=1, failed=0, records=50000, max_depth=2
Final tree breakdown: account=0/1, campaigns=4/4, max_depth=2
```

---

---

**Test Organization**

Reorganized the test directory into a clearer structure:

```
tests/
├── conftest.py              # Shared fixtures
├── factories/               # Reusable test helpers
│   ├── http.py              # MockHTTPSession with URL routing
│   └── responses.py         # Facebook API response factories
├── unit/                    # Pure logic tests (no I/O)
│   └── test_job_manager.py
├── integration/             # Mocked HTTP tests
│   └── test_job_manager.py
└── e2e/                     # Full connector tests with flowctl
    ├── test_snapshots.py    # Moved from tests/
    └── snapshots/           # Unchanged snapshot files
```

*Key changes:*
- Moved `test_snapshots.py` to `tests/e2e/` — these are end-to-end tests that run the full connector via flowctl
- Snapshot JSON files unchanged — confirms the refactoring introduced no behavioral changes to spec/discover output
- New `tests/unit/` for pure logic tests without I/O dependencies
- New `tests/integration/` for tests with mocked HTTP responses
- New `tests/factories/` with reusable test infrastructure

---

**Test Infrastructure**

*MockHTTPSession* (`tests/factories/http.py`):
- URL-pattern routing with regex matching
- HTTP method filtering (GET vs POST)
- Concurrency tracking for semaphore tests
- FIFO fallback queue for simple sequential tests

*Response factories* (`tests/factories/responses.py`):
- `job_submission_response()` — async job creation
- `job_status_response()` — polling status with configurable progress
- `insights_response()` — paginated insights data
- `entity_discovery_response()` — campaign/adset/ad ID discovery
- `data_limit_error_response()` — code 100, subcode 1487534 errors

---

**Test Coverage**

*Unit tests* (`tests/unit/test_job_manager.py`) validate core splitting logic:
- `InsightsJob.can_split()` for all scope/entity combinations
- `_binary_split()` for edge cases (empty, single, even, odd counts)
- `try_parse_facebook_api_error()` for data limit detection

*Integration tests* (`tests/integration/test_job_manager.py`) validate orchestration with mocked HTTP:

| Scenario | Behavior |
|----------|----------|
| Happy path | Submit → poll → fetch results |
| Pagination | Multi-page results correctly yielded |
| Job Failed status | Retry 3x then split |
| Data limit error | Immediate split, no retries |
| Binary split | 4 campaigns → 2 jobs of 2 |
| Hierarchy descent | Account → Campaign → AdSet → Ad |
| Atomic failure | Single ad raises `CannotSplitFurtherError` |
| Empty discovery | Graceful completion with 0 results |
| Semaphore limit | Concurrent jobs bounded correctly |
| Multi-account | Separate managers execute independently |

*E2E tests `(snapshot tests)`* (`tests/e2e/test_snapshots.py`) validate full connector behavior:
- Spec output matches expected schema
- Discover output matches expected catalog
- Snapshot files unchanged from before refactoring

---

**Files Changed**

*New files:*
- `source_facebook_marketing_native/insights/__init__.py`
- `source_facebook_marketing_native/insights/errors.py`
- `source_facebook_marketing_native/insights/types.py`
- `source_facebook_marketing_native/insights/metrics.py`
- `source_facebook_marketing_native/insights/manager.py`
- `tests/conftest.py`
- `tests/factories/http.py`
- `tests/factories/responses.py`
- `tests/unit/test_job_manager.py`
- `tests/integration/test_job_manager.py`
- `tests/README.md`

*Moved files:*
- `tests/test_snapshots.py` → `tests/e2e/test_snapshots.py`
- `tests/snapshots/` → `tests/e2e/snapshots/` (contents unchanged)

*Modified files:*
- `source_facebook_marketing_native/models.py` — Consolidate `BaseAdsInsights` into `FacebookInsightsResource`
- `source_facebook_marketing_native/api.py` — Integrate with job manager
- `source_facebook_marketing_native/resources.py` — Create per-account job managers

*Deleted files:*
- `source_facebook_marketing_native/job_manager.py` — Replaced by `insights/` module

---

**Testing**

```bash
poetry run pytest tests/ -v
# 28 passed
```

```bash
poetry run mypy source_facebook_marketing_native/insights/
# Success: no issues found in 5 source files
```

**Workflow steps:**

(How does one use this feature, and how has it changed)

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

- I've done a self review and went through a few iterations before landing here, so I think things are solid.
- I used Claude to produce much of this along with the unit, integration, and e2e tests structuring and test cases - I think things turned out pretty well.
- The new tests give me a good bit of confidence, but I also ran through a few manual tests with a real FB account using `flowctl preview` (local stack wasn't set up fully).
  - I set the number of days used in the backfill to a large number like `60` or `100+` days instead of 1 to force a split at least at the account level.
  - In my testing I only saw 3 jobs total (account level - failed 3x, and 2 campaign jobs that succeeded with 1 needing to retry once)

### CODE REVIEW

**Review by JustinASmith — COMMENTED** [review]
@Alex-Bair @nicolaslazo  - I had a few review items for myself I missed on Friday due to my late Friday evening brain fog.

File: source-facebook-marketing-native/source_facebook_marketing_native/insights/manager.py [hunk 1]
```diff
@@ -0,0 +1,819 @@
...
+        result_queue: asyncio.Queue[InsightRecord] = asyncio.Queue()
+        metrics = JobMetrics()
+        last_progress_log = 0
+
+        async def execute_with_semaphore(job: InsightsJob) -> JobOutcome:
+            """Wrapper to acquire semaphore before execution."""
+            async with self._semaphore:
+                return await self._execute_job(
+                    log, job, model, account_id, time_range, result_queue
+                )
```
Comment (Alex-Bair) [inline source-facebook-marketing-native/source_facebook_marketing_native/insights/manager.py]: I initially thought the `self._semaphore` use in `execute_with_semaphore` and the `while queue and len(pending) < MAX_CONCURRENT_JOBS` on line 608 were limiting the same thing, but they're not. 

`self._semaphore` is a `FacebookInsightsJobManager` limit shared across all `_process_with_splitting` invocations this job manager instance is managing. The check on line 608 is ensuring each `_process_with_splitting` invocation never has more than `MAX_CONCURRENT_JOBS` in `pending` (not necessarily running).

Using the same `MAX_CONCURRENT_JOBS` for each feels misleading - the semaphore does the actual limiting in all cases while line 608 makes sure no more than `MAX_CONCURRENT_JOBS` are attempting to run per `_process_with_splitting` invocation (which may be waiting to acquire the job-manager-wide semaphore, not actually running). For clarity, it'd be worth considering either renaming one of the constants or adding a comment somewhere to make that distinction clearer.

**Review by Alex-Bair — APPROVED** [review]
LGTM % a few comments + the items you noted.

File: source-facebook-marketing-native/source_facebook_marketing_native/insights/manager.py [hunk 2]
```diff
@@ -0,0 +1,819 @@
...
+        max_job_wait_time: timedelta = timedelta(hours=2),
+        max_retries: int = 3,
+    ):
+        self._http = http
+        self._base_url = base_url
+        self._log = log
+        self._account_id = account_id
+        self._poll_interval = poll_interval
+        self._max_job_wait_time = max_job_wait_time
+        self._max_retries = max_retries
```
[the remark this answers is not in the document]
Comment (JustinASmith) [reply source-facebook-marketing-native/source_facebook_marketing_native/insights/manager.py]: It's not.. the constant was used instead. I'll go through the constructor of the `FacebookInsightsJobManager` again and remove most of these to remove the configuration until it is needed outside of the job manager itself. Then use the constants that are defined.
